### DISCUSSION TITLE
What about .vbg

### DISCUSSION THREAD

**Discussion by MatTheCatz** [twinbasic/twinbasic#1716] [General]
I have a big project Vb6 with several .vbp (dll, activex, exe) in one .vbg. How can i do for open a .vbg with TwinBasic?

Actualy on your road map, the App object is completed, or when i call App.LogEvent, i have a error Logevent is not recognised. Why? Roadmap is not udpated?

**Comment by fafalone** [comment]
The roadmap is perspective plans and hasn't been updated since June; some things get rearranged or fall behind. Or in some cases, come ahead of schedule (for example, vbWatchDog is now available and optimization is partially implemented). The App object is not yet completed, but is planned for version 1.0. 

Current status of VBG support is not planned for 1.0, so unfortunately it's at least a few months out unless that changes (which wouldn't be unheard of, Wayne certainly takes input from the community on priorities).

But a VBG file is just a list of .VBP projects to open. You can open all of your VBPs in different instances of twinBASIC. Also, it's not implemented yet, but IIRC command line compilation is planned for 1.0 so likely sooner than direct VBG support... once that's available you'd be able to mark a sub `[ RunAfterBuild ]` and use it to shell the tB compiler on the other projects. The community may even be able to help with an addin to read VBG then shell the compiler. Or if the addin infrastructure adds support for loading projects it could load them sequentially.

**Comment by EduardoVB** [comment]
VBG support is a lot more than just opening individual files. It is to be able to run an ActiveX (OCX) project in source code and the associated client program (usually a test project) at once.
VB6 compiles the OCX in p-code and uses that for the test project.

Another important (at least for me) feature that is missing is Stop-EditCode-And-Continue.

**Comment by WaynePhillipsEA** [comment]
@EduardoVB for that particular use case, a temporary workaround is to add the test form that uses the control to the OCX project.  You can then use the 'PREVIEW' button from the test form designer to debug it.

**Comment by EduardoVB** [comment]
Yes, for the particular case of testing. 
Anyway the point was that project groups is mostly that, not just a bunch of single projects, but usually a main exe project and maybe an OCX project and also a couple of DLL projects, all related and that need to work together.

**Comment by fafalone** [comment]
I think the 'at once' part will get better once we can use .twinproj files like VBPs. That way when you save the file you can just reload it in another project with one click. Also a planned 1.0 feature last I checked. If they're compiled DLLs or OCXs they're already externally linked so you should be able to refresh the test project after a build.

**Comment by EduardoVB** [comment]
Then the programmer usually works most in the main exe, but sometimes need to fix something, add something or make changes in the OCX that may hold some specific controls, or in one of the DLLs that may have some specific financial math functions or other functions specific to that domain.
And from time to time makes a release, compiling all, the exe and all the dependencies.

I think that's a normal use case of project groups, besides the OCX/DLL testing purpose.

I mean: the main exe and the OCX and DLL are developed "at the same time".

**Comment by wqweto** [comment]
I thought that project groups become obsolete with the new package manager. I mean everything you reference as source package is included in current "group" i.e. every .twinproj is actually a .vbg when referencing source packages.

If this is not the case then probably more research is needed on how modern languages deal with multi-project solutions -- TypeScript, JS, python, rust, go, etc.

Btw, importing a .vbg file can produce a .twinproj with the startup project as current sources and every other one referenced as a "local" source package I guess.

**Comment by EduardoVB** [comment]
Can you develop the main project along with the packages at the same time? (and run all -the modified version- just pressing a key -F5-?)

**Comment by fafalone** [comment]
No, packages are read-only. That's why I was saying once external files are supported... then you might have to click a 'Reload' button first, but you wouldn't have to go through the whole process of removing and readding as you do currently for a package (I've suggest a 'update packages' button/menu option before, perhaps we could get that before 1.0 to help with this?)


As for other languages, full blown Visual Studio and Delphi also handle project groups fairly well, and essentially the same as VB6. So it's pretty much a solved problem, it's just there's a lot of other things that are higher priority for v1.0 in Wayne's view... I'd tend to agree, if we could get the other features already mentioned as likely for v1.0 (twinproj-like-VBP, command line compile).

**Comment by EduardoVB** [comment]
Semi-OT: I'm a little worried.
My perception is that in the last 8 months tB development slowed down a lot.
There are many important and also difficult things left to do just to be on par with VB6.
When? At this pace it seems they'll take several years.

**Comment by WaynePhillipsEA** [comment]
It's a fair comment.  Things did slow down for a while.  If you were following along you'd be aware that I moved home about three months ago, so the lead up to that, and the aftermath, certainly did slow down my output.  There was a also a family bereavement earlier in the year.  Not excuses... just trying to explain.

My commitment to this project is 100%, and output is increasing again now.  This project is certainly never standing still, so keep watching.

**Comment by EduardoVB** [comment]
I don't want to get into your personal life, but: maybe a vacation? (no need to answer, but I found myself that sometimes vacations are refreshing)
